src/dmd/expressionsem.d

```diff
@@ -7734,28 +7742,57 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
                         if (e2x.isLvalue())
                         {
-                            if (!e2x.type.implicitConvTo(e1x.type))
+                            if (sd.copyCtor)
                             {
-                                exp.error("conversion error from `%s` to `%s`",
-                                    e2x.type.toChars(), e1x.type.toChars());
-                                return setError();
+                                /* Rewrite as:
+                                 * e1 = init, e1.copyCtor(e2);
+                                 */
+                                Expression einit = new BlitExp(exp.loc, exp.e1, getInitExp(sd, exp.loc, sc, t1));
+                                einit.type = e1x.type;
+
+                                if (!einit)
+                                    return setError();
+
+                                Expression e;
+                                e = new DotIdExp(exp.loc, e1x, Id.copyCtor);
+                                e = new CallExp(exp.loc, e, e2x);
+                                e = new CommaExp(exp.loc, einit, e);
+
+                                //printf("e: %s\n", e.toChars());
+
+                                /* If semantic is performed correctly on e
+                                 * there is a copy constructor overload to be used.
+                                 * Otherwise, implicit copying may be used
+                                 */
+                                result = e.expressionSemantic(sc);
+                                if (result.op != TOK.error)
+                                    return;
+                            }
+                            else
+                            {
+                                if (!e2x.type.implicitConvTo(e1x.type))
+                                {
+                                    exp.error("conversion error from `%s` to `%s`",
```

Comment:
I imagine this error message would be completely baffling to a user.

Reply:
It was there before. It has nothing to do with this PR